contracts/overseer/src/contract.rs

```diff
@@ -287,15 +311,97 @@ pub fn update_whitelist(
     ]))
 }
 
+fn update_rate(old_rate: Decimal256, rate_change: Decimal256, direction: bool) -> Decimal256 {
+    if direction {
+        old_rate + rate_change
+    } else if old_rate > rate_change {
+        old_rate - rate_change
+    } else {
+        old_rate
+    }
+}
+
 pub fn execute_epoch_operations(deps: DepsMut, env: Env) -> Result<Response, ContractError> {
-    let config: Config = read_config(deps.storage)?;
+    let mut config: Config = read_config(deps.storage)?;
+    let dynrate_config: DynrateConfig = read_dynrate_config(deps.storage)?;
     let state: EpochState = read_epoch_state(deps.storage)?;
+    let dynrate_state: DynrateState = read_dynrate_state(deps.storage)?;
+
+    // check whether call came to early
     if env.block.height < state.last_executed_height + config.epoch_period {
         return Err(ContractError::EpochNotPassed(state.last_executed_height));
     }
 
-    // # of blocks from the last executed height
-    let blocks = Uint256::from(env.block.height - state.last_executed_height);
+    // retrieve interest buffer
+    let mut messages: Vec<CosmosMsg> = vec![];
+    let mut interest_buffer = query_balance(
+        deps.as_ref(),
+        env.contract.address.clone(),
+        config.stable_denom.to_string(),
+    )?;
+
+    // check whether its time to re-evaluate rate
+    if !dynrate_state.prev_yield_reserve.is_zero()
+        && env.block.height > dynrate_state.last_executed_height + dynrate_config.dyn_rate_epoch
+    {
+        let blks = Uint256::from(env.block.height - dynrate_state.last_executed_height);
+
+        // yield reserve amt
+        let yield_reserve = Decimal256::from_uint256(interest_buffer);
+
+        // direction of rate change
+        let up_down = yield_reserve > dynrate_state.prev_yield_reserve;
+
+        // normalized change in yr during dyn_rate_epoch
+        let yield_reserve_change = (if up_down {
+            yield_reserve - dynrate_state.prev_yield_reserve
+        } else {
+            dynrate_state.prev_yield_reserve - yield_reserve
+        }) / dynrate_state.prev_yield_reserve;
+
+        // yr change per block
+        let mut yield_reserve_change_pb = yield_reserve_change / Decimal256::from_uint256(blks);
+
+        // increase expectation pb
+        let increase_expectation_pb =
+            dynrate_config.dyn_rate_yr_increase_expectation / Decimal256::from_uint256(blks);
+
+        // consider increase expectation
+        yield_reserve_change_pb = if yield_reserve_change_pb > increase_expectation_pb {
+            yield_reserve_change_pb - increase_expectation_pb
+        } else {
+            yield_reserve_change_pb
+        };
+
+        // change exceeded rate threshold, need to update variable rate
+        // recalc values from confing to a per block
+        let year = Decimal256::from_str("4656810").unwrap();
```

Comment:
lets make this a constant and not use magic numbers

Reply:
good point, done